lib/api/CDataFrameAnalyzer.cc

```diff
@@ -6,15 +6,133 @@
 #include <api/CDataFrameAnalyzer.h>
 
 #include <core/CContainerPrinter.h>
+#include <core/CFloatStorage.h>
 #include <core/CLogger.h>
 
+#include <maths/CTools.h>
+
+#include <limits>
+
 namespace ml {
 namespace api {
+namespace {
+double truncateToFloatRange(double value) {
+    double largest{static_cast<double>(std::numeric_limits<float>::max())};
+    return maths::CTools::truncate(value, -largest, largest);
+}
+
+const std::string CONTROL_MESSAGE_FIELD_NAME{"."};
+// Control message types:
+const char RUN_ANALYSIS_CONTROL_MESSAGE_FIELD_VALUE{'r'};
+}
+
+// TODO memory calculations plus choose data frame storage strategy.
+CDataFrameAnalyzer::CDataFrameAnalyzer(CDataFrameAnalysisSpecification analysisSpecification)
+    : m_AnalysisSpecification{std::move(analysisSpecification)},
+      m_DataFrame{core::makeMainStorageDataFrame(m_AnalysisSpecification.cols())} {
+}
+
+bool CDataFrameAnalyzer::usingControlMessages() const {
+    return m_ControlFieldValue >= 0;
+}
 
 bool CDataFrameAnalyzer::handleRecord(const TStrVec& fieldNames, const TStrVec& fieldValues) {
-    LOG_INFO(<< core::CContainerPrinter::print(fieldNames));
-    LOG_INFO(<< core::CContainerPrinter::print(fieldValues));
+
+    // Control messages are signified by a dot in the field name. This supports:
+    //   - using the last field for a control message,
+    //   - missing.
+    //
+    // Note if the the control message field name is missing the analysis must
+    // be triggered to run by calling run explicitly.
+
+    bool result{true};
+    if (this->readyToReceiveControlMessages() == false) {
+        result = this->prepareToReceiveControlMessages(fieldNames);
+    }
+
+    if (this->isControlMessage(fieldValues)) {
+        return this->handleControlMessage(fieldValues);
+    }
+
+    this->addRowToDataFrame(fieldValues);
+
+    return result;
+}
+
+void CDataFrameAnalyzer::run() {
+    // TODO
+}
+
+bool CDataFrameAnalyzer::readyToReceiveControlMessages() const {
+    return m_ControlFieldValue != CONTROL_FIELD_UNSET;
+}
+
+bool CDataFrameAnalyzer::prepareToReceiveControlMessages(const TStrVec& fieldNames) {
+    if (fieldNames.empty() || fieldNames.back() != CONTROL_MESSAGE_FIELD_NAME) {
+        m_ControlFieldValue = CONTROL_FIELD_MISSING;
+        m_BeginDataFieldValues = 0;
+        m_EndDataFieldValues = static_cast<std::ptrdiff_t>(fieldNames.size());
+    } else {
+        m_ControlFieldValue = static_cast<std::ptrdiff_t>(fieldNames.size() - 1);
+        m_BeginDataFieldValues = 0;
+        m_EndDataFieldValues = m_ControlFieldValue;
+        auto pos = std::find(fieldNames.begin(), fieldNames.end(), CONTROL_MESSAGE_FIELD_NAME);
+        if (pos != fieldNames.end()) {
+            LOG_ERROR(<< "Unexpected possible control field: ignoring");
+            return false;
+        }
+    }
     return true;
 }
+
+bool CDataFrameAnalyzer::isControlMessage(const TStrVec& fieldValues) const {
+    return m_ControlFieldValue != CONTROL_FIELD_MISSING &&
```

Comment:
Shouldn't this be checking `m_ControlFieldValue >= 0 && fieldValues[m_ControlFieldValue].size() > 0`?

Reply:
It looks like this has been changed to `m_ControlFieldValue >= CONTROL_FIELD_MISSING`, which could lead to a SEGV.  It either needs to be `>= 0` or `!= CONTROL_FIELD_MISSING && != CONTROL_FIELD_UNSET`.

Also, it might be worth renaming `m_ControlFieldValue` to `m_ControlFieldIndex`, as "value" is easy to confuse with the control message itself.